File: Blockchain/node-api/api/routes/doctor.js
const express = require("express");
const { v4: uuidv4 } = require("uuid");
const { contract } = require("../contract");
const isAuthorize = require("../util/isAuthorizedAdmin");
routes = express.Router();

routes.put("/report/addcomment", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    ["UpdateReportComment", req.body.report_id, "0", req.body.comment],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        if (payload) {
          res.status(200).json({
            message: `Successfuly added comment to report ${req.body.report_id}`,
          });
        } else {
          res.status(400).json({
            message: `Something went wrong!`,
          });
        }
      }
    }
  );
});

routes.put("/report/updateStatus", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    ["UpdateReportStatus", req.body.report_id, req.body.status],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        if (payload) {
          res.status(200).json({
            message: `Successfuly update the status of report ${req.body.report_id}`,
          });
        } else {
          res.status(400).json({
            message: `Something went wrong!`,
          });
        }
      }
    }
  );
});

routes.post("/report/reftest", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    [
      "RefTest",
      req.body.patient_id,
      req.body.report_id,
      req.body.labID,
      req.body.name,
      req.body.ref_doctor,
      req.body.type_of_test,
    ],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json({
          message: `Successfuly created test of type ${req.body.type_of_test}`,
          test_id: payload.toString(),
        });
      }
    }
  );
});
routes.post("/report/reftreatment", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    [
      "RefTreatment",
      req.body.patient_id,
      req.body.report_id,
      req.body.ref_doctor,
      req.body.name,
    ],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json({
          message: `Successfuly created treatment of name ${req.body.name}`,
          test_id: payload.toString(),
        });
      }
    }
  );
});

routes.post("/report/presdrugs", (req, res) => {
  console.log(req.body);
  contract(
    req.body.clientId,
    "INVOKE",
    [
      "PrescribeDrugs",
      req.body.patient_id,
      req.body.reportID,
      req.body.refDoctor,
      req.body.pharamacyID,
      req.body.drugs_note_media,
    ],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json({
          message: `Successfuly prescribed drugs`,
          drug_id: payload.toString(),
        });
      }
    }
  );
});

routes.put("/treatment/updateStatus", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    [
      "UpdateTreatmentStatus",
      req.body.treatment_id,
      "0",
      req.body.comment,
      req.body.status,
    ],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json({
          message: `Successfuly updated the status on treatment`,
        });
      }
    }
  );
});

routes.put("/treatment/updateMediaFile", (req, res) => {
  const mediaFileLocation = req.body.mediaFileLocation;
  contract(
    req.body.clientId,
    "INVOKE",
    [
      "UpdateMediaFilesTreatment",
      req.body.treatment_id,
      req.body.media_no,
      mediaFileLocation,
    ],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json({
          message: `Successfuly updated the media files on treatment`,
        });
      }
    }
  );
});

routes.put("/treatment/addmediafile", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    [
      "AddMediaToTreatment",
      req.body.treatment_id,
      req.body.supervisor,
      req.body.no_of_mediafile,
    ],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json(JSON.parse(payload));
      }
    }
  );
});

routes.put("/test/updateMediaFile", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    [
      "UpdateMediaFilesTest",
      req.body.testID,
      req.body.media_no,
      req.body.mediaFileLocation,
    ],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json({
          message: `Successfuly updated the media files on test`,
        });
      }
    }
  );
});

routes.put("/test/UpdateStatusTest", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    ["UpdateStatusTest", req.body.testID, req.body.status],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json({
          message: `Successfuly updated the status on test`,
        });
      }
    }
  );
});

routes.put("/test/UpdateResultTest", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    ["UpdateResultTest", req.body.testID, req.body.result],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json({
          message: `Successfuly updated the result on test`,
        });
      }
    }
  );
});

// createDoctor route
routes.post("/doctor", async (req, res) => {
  const isAuthorizedToAccess = await isAuthorize(req.headers.authorization);
  if (isAuthorizedToAccess) {
    contract(
      req.body.clientId,
      "INVOKE",
      [
        "CreateDoctor",
        uuidv4(),
        req.body.name,
        req.body.email,
        req.body.specialty,
        req.body.licenseNo,
        req.body.phoneNumber,
        req.body.address,
      ],
      (err, payload) => {
        if (err) {
          res.status(500).json(err);
        } else {
          res.status(200).json({
            message: `Successfully created doctor ${req.body.name}`,
            data: payload,
          });
        }
      }
    );
  } else {
    res.status(500).json({ message: "unauthorized!" });
  }
});

// getAllDoctors route
routes.get("/doctors", (req, res) => {
  contract(req.query.clientId, "QUERY", ["GetAllDoctors"], (err, payload) => {
    if (err) {
      res.status(500).json(err);
    } else {
      const doctors = JSON.parse(payload);
      res.status(200).json(doctors);
    }
  });
});

routes.get("/doctors-count", (req, res) => {
  contract(req.query.clientId, "QUERY", ["GetAllDoctors"], (err, payload) => {
    if (err) {
      console.log(err);
      res.status(500).json(err);
    } else {
      const hospitals = JSON.parse(payload);
      res.status(200).json(hospitals.length);
    }
  });
});

routes.get("/doctor", (req, res) => {
  const doctorID = req.headers.doctorid;
  contract(
    req.query.clientId,
    "QUERY",
    ["GetDoctorByID", doctorID],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        const doctor = JSON.parse(payload);
        res.status(200).json(doctor);
      }
    }
  );
});

module.exports = routes;

File: Blockchain/Chaincode/doctor.go
package main

import (
	"encoding/json"
	"fmt"
	. "fmt"
	"strconv"
	"time"
)

const (
	CREATED = "CREATED"
	UPDATED = "UPDATED"
)

type ReportOutput struct {
	Result []Report `json:"reports"`
}
type TreatmentOutput struct {
	Result []Treatment `json:"result"`
}
type TestOutput struct {
	Result []Test `json:"result"`
}

func (c *Chaincode) RefTest(ctx CustomTransactionContextInterface, patientID, reportID, labID, name, refDoctor string, typeoftest int) (string, error) {
	// if ctx.GetData() == nil {
	// 	return "", Errorf("Report with ID %v doesn't exists", reportID)
	// }

	// var report Report
	// json.Unmarshal(ctx.GetData(), &report)
	// if report.Doctor != doc from certs {
	// 	return Errorf("Cannot prescribe drug from the patient")
	// }
	// if ok := c.checkConsent(ctx, report.PatientID, refDoctor); !ok {
	// 	return "", Errorf("No consent from the patient")
	// }

	id := TESTS + getSafeRandomString(ctx.GetStub())
	test := Test{
		DocTyp:            TESTS,
		ReportID:          reportID,
		PatientID:         patientID,
		LabID:             labID,
		ID:                id,
		Name:              name,
		RefDoctor:         refDoctor,
		Status:            0,
		MediaFileLocation: []string{},
		CreateTime:        time.Now().Unix(),
		UpdateTime:        time.Now().Unix(),
	}
	if typeoftest == 1 {
		test.TypeOfT = 1
	}
	testAsByte, _ := json.Marshal((test))
	return test.ID, ctx.GetStub().PutState(id, testAsByte)
}

func (c *Chaincode) RefTreatment(ctx CustomTransactionContextInterface, patientID, reportID, refDoctor, name string) (string, error) {
	// if ctx.GetData() == nil {
	// 	return "", Errorf("Report with ID %v doesn't exists", reportID)
	// }

	// var report Report
	// json.Unmarshal(ctx.GetData(), &report)
	// if report.Doctor != doc from certs {
	// 	return Errorf("Cannot prescribe drug from the patient")
	// }
	// if ok := c.checkConsent(ctx, report.PatientID, refDoctor); !ok {
	// 	return "", Errorf("No consent from the patient")
	// }
	id := TREATMENT + getSafeRandomString(ctx.GetStub())
	treatment := Treatment{
		DocTyp:            TREATMENT,
		ReportID:          reportID,
		ID:                id,
		RefDoctor:         refDoctor,
		Name:              name,
		Comments:          make(map[string]string),
		Status:            0,
		MediaFileLocation: []string{},
		CreateTime:        time.Now().Unix(),
		UpdateTime:        time.Now().Unix(),
		// PatientID:         report.PatientID,
		PatientID: patientID,
	}
	treatmentAsByte, _ := json.Marshal(treatment)
	return treatment.ID, ctx.GetStub().PutState(treatment.ID, treatmentAsByte)
}

func (c *Chaincode) PrescribeDrugs(ctx CustomTransactionContextInterface, patientID, reportID, refDoctor, pharamacyID, Drugs_note_media string) (string, error) {

	id := DRUGS + getSafeRandomString(ctx.GetStub())
	drugs := Drugs{
		DocTyp:           DRUGS,
		ReportID:         reportID,
		PatientID:        patientID,
		PharamacyID:      pharamacyID,
		ID:               id,
		RefDoctor:        refDoctor,
		Drugs_note_media: Drugs_note_media,
		CreateTime:       time.Now().Unix(),
		UpdateTime:       time.Now().Unix(),
	}
	drugsAsByte, _ := json.Marshal(drugs)
	return drugs.ID, ctx.GetStub().PutState(drugs.ID, drugsAsByte)
}

func (c *Chaincode) AddCommentsToReport(ctx CustomTransactionContextInterface, reportID, comment, refDoctor string) error {
	// if ctx.GetData() == nil {
	// 	return Errorf("Report with ID %v doesn't exists", reportID)
	// }

	var report Report
	json.Unmarshal(ctx.GetData(), &report)
	// if report.Doctor != doc from certs {
	// 	return Errorf("Cannot prescribe drug from the patient")
	// }
	if ok := c.checkConsent(ctx, report.PatientID, refDoctor); !ok {
		return Errorf("No consent from the patient")
	}
	timeNow := time.Now().Unix()
	stringTime := strconv.FormatInt(timeNow, 10)
	report.UpdateTime = time.Now().Unix()
	report.Comments[stringTime] = comment
	reportAsByte, _ := json.Marshal(report)

	return ctx.GetStub().PutState(report.ID, reportAsByte)
}

func (c *Chaincode) AddCommentsToTreatment(ctx CustomTransactionContextInterface, treatmentID, superviosr, comment string) error {
	if ctx.GetData() == nil {
		return Errorf("Treatment with ID %v doesn't exists", treatmentID)
	}
	var treatment Treatment
	json.Unmarshal(ctx.GetData(), &treatment)
	if ok := c.checkConsent(ctx, treatment.PatientID, superviosr); !ok {
		return Errorf("no consent from the patient")
	}
	if treatment.Status == 2 {
		return Errorf("Treatment is already completed")
	}
	timeNow := time.Now().Unix()
	stringTime := strconv.FormatInt(timeNow, 10)
	treatment.Comments[stringTime] = comment

	treatment.UpdateTime = time.Now().Unix()
	treatmentAsByte, _ := json.Marshal(treatment)
	return ctx.GetStub().PutState(treatment.ID, treatmentAsByte)
}

func (c *Chaincode) AddMediaToTreatment(ctx CustomTransactionContextInterface, treatmentID, superviosr string, numberOfMfile int) ([]string, error) {
	if ctx.GetData() == nil {
		return []string{}, Errorf("Treatment with ID %v doesn't exists", treatmentID)
	}
	var treatment Treatment
	json.Unmarshal(ctx.GetData(), &treatment)
	if ok := c.checkConsent(ctx, treatment.PatientID, superviosr); !ok {
		return []string{}, Errorf("No consent from the patient")
	}
	for i := 0; i < numberOfMfile; i++ {
		id := TREATMENT + "media" + getSafeRandomString(ctx.GetStub()) + strconv.Itoa(i)
		treatment.MediaFileLocation = append(treatment.MediaFileLocation, id)
	}
	treatment.UpdateTime = time.Now().Unix()
	treatmentAsByte, _ := json.Marshal(treatment)
	return treatment.MediaFileLocation, ctx.GetStub().PutState(treatment.ID, treatmentAsByte)
}

// Create a new doctor
func (c *Chaincode) CreateDoctor(ctx CustomTransactionContextInterface, id string, name string, email string, specialty string, licenseNo string, phoneNumber string, address string) error {
	// Check if the doctor with the given ID already exists
	exists, err := c.doctorExists(ctx, id)
	if err != nil {
		return err
	}
	if exists {
		return fmt.Errorf("the doctor with ID %s already exists", id)
	}

	// Create a new doctor struct
	doctor := Doctor{
		DocTyp:      DOCTOR,
		ID:          id,
		Name:        name,
		Email:       email,
		Specialty:   specialty,
		LicenseNo:   licenseNo,
		PhoneNumber: phoneNumber,
		Address:     address,
		CreatedAt:   time.Now().Unix(),
		UpdatedAt:   time.Now().Unix(),
	}

	// Convert the doctor struct to JSON format
	doctorJSON, err := json.Marshal(doctor)
	if err != nil {
		return err
	}

	// Save the doctor to the world state
	err = ctx.GetStub().PutState(id, doctorJSON)
	if err != nil {
		return err
	}

	return nil
}

// doctorExists returns true if the doctor with given ID exists in the ledger
func (c *Chaincode) doctorExists(ctx CustomTransactionContextInterface, id string) (bool, error) {
	doctorJSON, err := ctx.GetStub().GetState(id)
	if err != nil {
		return false, err
	}
	return doctorJSON != nil, nil
}

// getAllDoctors returns all doctors in the ledger
func (s *Chaincode) GetAllDoctors(ctx CustomTransactionContextInterface) ([]*Doctor, error) {
	// Create a new query string to get all DOCTOR documents
	queryString := fmt.Sprintf(`{"selector":{"docTyp":"%s"}}`, DOCTOR)

	// Create a new query iterator using the query string
	queryResults, err := ctx.GetStub().GetQueryResult(queryString)
	if err != nil {
		return nil, fmt.Errorf("failed to execute query: %v", err)
	}
	defer queryResults.Close()

	// Create a slice to hold the results
	var doctors []*Doctor

	// Iterate over the query results and deserialize each document
	for queryResults.HasNext() {
		queryResult, err := queryResults.Next()
		if err != nil {
			return nil, fmt.Errorf("failed to get next query result: %v", err)
		}

		// Deserialize the document into a Doctor struct
		var doctor Doctor
		err = json.Unmarshal(queryResult.Value, &doctor)
		if err != nil {
			return nil, fmt.Errorf("failed to deserialize doctor: %v", err)
		}

		// Add the doctor to the results slice
		doctors = append(doctors, &doctor)
	}

	return doctors, nil

}

func (c *Chaincode) GetDoctorByID(ctx CustomTransactionContextInterface, doctorID string) (*Doctor, error) {
	// Create a new query string to get the DOCTOR document for the given reportID
	queryString := fmt.Sprintf(`{"selector":{"docTyp":"%s", "id": "%s"}}`, DOCTOR, doctorID)

	// Create a new query iterator using the query string
	queryResults, err := ctx.GetStub().GetQueryResult(queryString)
	if err != nil {
		return nil, fmt.Errorf("failed to execute query: %v", err)
	}
	defer queryResults.Close()

	// Iterate over the query results and deserialize the document
	if queryResults.HasNext() {
		queryResult, err := queryResults.Next()
		if err != nil {
			return nil, fmt.Errorf("failed to get next query result: %v", err)
		}

		// Deserialize the document into a Doctor struct
		var doctor Doctor
		err = json.Unmarshal(queryResult.Value, &doctor)
		if err != nil {
			return nil, fmt.Errorf("failed to deserialize doctor: %v", err)
		}

		return &doctor, nil
	}

	return nil, nil // return nil if no matching doctor found
}
